# Cálculo de raíces cuadradas

Utilizaremos el método de la sucesión convergente para calcular la raíz cuadrada de un número $a\in\mathbb{R}$ positivo:

\begin{gather}
 x_0 \in \mathbb{R}: \; x_0^2\approx a \\
 x_{n+1} = \frac{1}{2}\left( x_n + \frac{a}{x_n} \right)\\
 (\text{Se sabe que}: \lim_{n\to\infty} x_n = \sqrt{a})
\end{gather}

In [1]:
from fractions import Fraction as F

Este módulo es útil para hacer operaciones con fracciones.

In [5]:
x=F(2,4) # 1/2
x

Fraction(1, 2)

Una de sus gracias es que las expresa en forma irreducible.

> **Nota**: Una fracción $\frac{a}{b}$ es irreducible cuando $a$ y $b$ son primos relativos (es decir, el máximo común divisor de $a$ y $b$ es 1).

Además, podemos obtener la representación decimal (_aproximada_) de esas fracciones.

In [6]:
float(x)

0.5

In [8]:
float(F(2,3))

0.6666666666666666

Vamos a definir una función, de forma recursiva, que nos permita obtener el término $x_n$ del listado de números que converge a la raíz de $a$. Para ello, vamos a ejemplificar la definición y uso de las funciones en [`Python 3`](https://www.python.org/).

In [9]:
# Vamos a hacer una función que, dados dos números, calcule su suma
def raul(a,b): #El nombre de la función es arbitrario
    return a+b

raul(2,3)

5

In [10]:
def memo(papitas,refresco): #También el nombre de los argumentos es arbitrario (aunque se prefiere algo que )
    return papitas*refresco # todavía no haya sido utilizado).

memo(5,2)

10

In [23]:
def raiz(a,x0,n):
    """
    a = número cuya raíz me interesa
    x0 = aproximación inicial de la raíz de a (se recomienda un entero o fracción cuyo cuadrado sea casi a)
    n = número natural para el que calcularemos x1, x2,..., xn
    """
    y=x0
    for i in range(n):
        print(f'X({i})={y}\n     ={float(y)}\n')
        y=F(1,2)*(y+F(a,y)) # Esto codifica la fórmula recursiva definida al inicio
    print(f'X({n})={y}\n     ={float(y)}')
    return y

In [25]:
y=raiz(10,3,5)

X(0)=3
     =3.0

X(1)=19/6
     =3.1666666666666665

X(2)=721/228
     =3.162280701754386

X(3)=1039681/328776
     =3.162277660169842

X(4)=2161873163521/683644320912
     =3.1622776601683795

X(5)=9347391150304592810234881/2955904621546382351702304
     =3.1622776601683795


## <font color='crimson'> Participantes </font>

* José Luis Carrizosa
* Omar Santiago Martínez González
* Aaron Fernando Del Valle Reyes

La aproximación obtenida para la raíz de 10 es tal, que el cuadrado _se pasa de 10_ en:

In [31]:
e=y**2-10
print(e)

1/8737372131679261877757908894145308222766638908416


In [32]:
float(e)

1.1445088808502047e-49

Este número está expresado en notación científica (computacionalmente):

$$ \text{1.1445088808502047e-49} = 1.1445088808502047 \times 10^{-49} = \\
0.00000000000000000000000000000000000000000000000011445088808502047
$$